### PR TITLE
Danyangliu/optimize dkn

### PR DESCRIPTION
**Pull request by danyang-liu**
### Description

Previous DKN using sparse operation and for loop to deal with the variable length of user click history, but in this way, training efficiency is very low, it takes to much time when training on big dataset. I padding user click history, change it to tensor operation. The time training on MIND whole dataset from about 13h to about 1.7h on my machine.

previous sparse operation too slow, it is needed to be optimized.

### Related Issues

https://github.com/microsoft/recommenders/issues/1145

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`.

### CODE REVIEW

File: reco_utils/recommender/deeprec/models/dkn.py [hunk 1]
```diff
@@ -176,7 +177,14 @@ def _build_dkn(self):
                 curr_hidden_nn_layer = tf.nn.xw_plus_b(
                     hidden_nn_layers[layer_idx], curr_w_nn_layer, curr_b_nn_layer
                 )
-                scope = "nn_part" + str(idx)
+                if hparams.enable_BN is True:
```
Comment (Leavingseason) [inline reco_utils/recommender/deeprec/models/dkn.py:180]: has you tested if enable_BN set to True/False, the model will go normally?

**Comment by danyang-liu** [comment]
> Thanks Danyang! This is great. Please inform us after you finish passing all the local test cases, plus one more case I mentioned for BatchNorm operation.

I finished running test, related test all passed. And I tested batch norm before, the model works normally.

File: examples/00_quick_start/dkn_MIND_dataset.ipynb [hunk 2]
```diff
@@ -328,8 +328,12 @@
     "| DKN | 0.6436 | 0.3128 | 0.3371 | 0.3908|\n",
```
Comment (Leavingseason) [inline examples/00_quick_start/dkn_MIND_dataset.ipynb:328]: Please remove the DeepFM's results from this table

**Comment by danyang-liu** [comment]
I add a quick for MINDsmall dataset.

**Comment by danyang-liu** [comment]
> this looks good, one issue:
> can you change the param `his_size` to `history_size`? The abbreviation is a different word making the variable sound odd. =)

Thanks! I changed it to history_size

File: examples/00_quick_start/dkn_MIND_dataset.ipynb [hunk 3]
```diff
@@ -30,13 +30,13 @@
     "- training / validation / test files: each line in these files represents one instance. Impressionid is used to evaluate performance within an impression session, so it is only used when evaluating, you can set it to 0 for training data. The format is : <br> \n",
     "`[label] [userid] [CandidateNews]%[impressionid] `<br> \n",
     "e.g., `1 train_U1 N1%0` <br> \n",
-    "- user history file: each line in this file represents a users' click history, the format is : <br> \n",
+    "- user history file: each line in this file represents a users' click history. You need to set his_size parameter in config file, which is the max number of user's click history we use. We will automatically keep the last his_size number of user click history, if user's click history is more than his_size, and we will automatically padding 0 if user's click history less than his_size. the format is : <br> \n",
```
Comment (anargyri) [inline examples/00_quick_start/dkn_MIND_dataset.ipynb:33]: Maybe calling this variable `history_size` would be more informative.

Comment (danyang-liu) [reply examples/00_quick_start/dkn_MIND_dataset.ipynb]: yes, I changed it to history_size
